stem/solver.py

```diff
@@ -1,10 +1,488 @@
+    displacement_absolute_tolerance: float = 1e-9
+    water_pressure_relative_tolerance: float = 1e-4
+    water_pressure_absolute_tolerance: float = 1e-9
+
+    @property
+    def convergence_criterion(self):
+        """
+        Property for returning the type of the displacement and water pressure convergence criterion
+
+        Returns:
+            - str: The type of the displacement and water pressure convergence criterion
+        """
+        return "displacement_and_water_pressure_criterion"
+
+
+@dataclass
+class NewmarkScheme(SchemeABC):
+    """
+    Class containing information about the Newmark scheme
+
+    Inheritance:
+        - :class:`SchemeABC`
+
+    Attributes:
+        - newmark_beta (float): The beta parameter of the Newmark scheme. Default value is 0.25.
+        - newmark_gamma (float): The gamma parameter of the Newmark scheme. Default value is 0.5.
+        - newmark_theta (float): The theta parameter of the Newmark scheme, which is used for water pressure. Default\
+            value is 0.5.
+    """
+    newmark_beta: float = 0.25
+    newmark_gamma: float = 0.5
+    newmark_theta: float = 0.5
+
+    @property
+    def scheme_type(self):
+        """
+        Property for returning the type of the Newmark scheme
+
+        Returns:
+            - str: The type of the newmark scheme
+        """
+        return "newmark"
+
+
+@dataclass
+class BackwardEulerScheme(SchemeABC):
+    """
+    Class containing information about the backward Euler scheme
+
+    Inheritance:
+        - :class:`SchemeABC`
     """
-    Class containing information about the builder, solver and strategy
+
+    @property
+    def scheme_type(self):
+        """
+        Property for returning the type of the backward Euler scheme
+
+        Returns:
+            - str: The type of the backward Euler scheme
+        """
+        return "backward_euler"
+
+
+class SolutionType(Enum):
+    """
+    Enum class containing the solution types
 
     Attributes:
-        - settings (dict): dictionary containing the solver settings
+        - QUASI_STATIC (int): quasi-static solution type
+        - K0_PROCEDURE (int): K0-procedure solution type
+        - DYNAMIC (int): dynamic solution type
+
+    """
+    QUASI_STATIC = 1
+    K0_PROCEDURE = 2
+    DYNAMIC = 3
+
+
+@dataclass
+class StrategyTypeABC(ABC):
+    """
+    Abstract class for the strategy type
+
+    Attributes:
+        - max_iterations (int): maximum number of iterations allowed, if this number is reached, the time step size is\
+            decreased and the algorithm is restarted. Default value is 15.
+        - min_iterations (int): minimum number of iterations, below this number, the time step size is increased.\
+            Default value is 6.
+        - number_cycles (int): number of allowed cycles of decreasing the time step size until the algorithm is stopped.\
+            Default value is 100.
+
+    """
+    max_iterations: int = 15
```

Comment:
it is not common to initilise variables in the abstract class
this is where it becomes tricky with inheretance (imo)